codeforamerica/vita-min
# Change GYR filing years to calculate valid filing years precisely

## Link to pivotal/JIRA issue
- https://codeforamerica.atlassian.net/browse/GYR1-478
## Is PM acceptance required?
- Yes - don't merge until JIRA issue is accepted!
## What was done?
- Changed the implementation of `MultiTenantService#filing_years` (for GYR) to calculate what tax years can still have returns filed according to IRS rules: i.e. 
  - have a start date that is in the past, and
  - have a due date less than 3 years in the past
- This implies that we no longer solely use `current_tax_year` to calculate filing years, and I propose that we might be able to get rid of it entirely - I've added a topic to computer gardening to discuss this PR and possible further changes.
## How to test?
- I fixed all the unit tests that this change broke, and added appropriate Timecop freezes in a number of places where we were testing only by setting `current_tax_year`
- This can result in fewer than 3 previous years being shown in the hub's `create_client` form, and in the backtaxes form in GYR.
- Additional manual testing should be done to confirm that all the flows work!
- Lastly, confirm that the homepage shows 2021-2023 - the simple-sounding original goal of this story.
## Screenshots (for visual changes)
- Before
<img width="663" alt="image" src="https://github.com/user-attachments/assets/cb7d56bb-d166-41b5-a2b5-03536582c87f">

- After
<img width="634" alt="image" src="https://github.com/user-attachments/assets/6ff76ba2-2664-4006-8b6f-beb35929cfa1">
